src/mono/mono/metadata/assembly.c

```diff
@@ -1470,44 +1471,46 @@ bundled_assembly_match (const char *bundled_name, const char *name)
 }
 
 static MonoImage *
-open_from_bundle_internal (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status, gboolean is_satellite)
+open_from_bundle_internal (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status)
 {
-	if (!bundles)
+	if (!bundles && !mono_bundled_resource_get_bundle_contains_assemblies ())
```

Comment:
what if instead of having 2 different way of looking up bundled assemblies in `open_from_bundle_internal`, we instead changed `mono_register_bundled_assemblies` to insert all its `MonoBundledAssembly` objects into the bundled-resource hash table, and then called `mono_bundled_resource_get_bundled_resource`?  So we would just have a single mechanism

Reply:
We can certainly do that if we believe we won't cause to much performance impact on existing API, adding things into the hash table for each registered assembly will come at a cost, but on the other hand lookups will be much more effective.

Reply:
I'm not sure there's any users of the old bundling API using the new dotnet/runtime mono, to be honest, aside from wasi+wasm.  And they would presumably move to the new API for all cases.

XA and XI don't use the bundling APIs. they do their own bundling and just add loader hooks to intercept assembly loading requests

Reply:
Its used on some other ports as well, but those are in our control so we could migrate to new API's going forward. If we don't think it will be used, then maybe we should deprecate the API's and make a note to use new bundling API's and register assemblies and managed assemblies into hash table.